_posts/2019-04-16-preservation.md

```diff
@@ -0,0 +1,358 @@
+
+1. The manifest files act as a pessimistic lock on the system.
+   * If we have to wait until everything is "done" uploading to update those
+      manifests, then we need to have some way to tell when we're "done." To do
+      this, in Ruby at least, either the uploads must be done sequentially on a
+      single system or some sort of external locking mechanism must be used to
+      keep track of current upload state.
+   * If instead of waiting we simply update the manifest when a file is
+      uploaded, then it means we need to have the ability to lock the file itself
+      in place to avoid writes from multiple processes running into each other.
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the manifest will be
+   corrupted by a race condition:
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved in its spot in the bag, updating manifest.
+   5. Initial preservation ends. Updates manifest, but has out-of-date checksum
+      information.
+
+#### OCFL
+
+Implementation notes for OCFL recommend things such as pair-trees for storing
+objects in the object root. I'm going to treat that as an implementation detail
+and pretend we're using a file system that can handle an infinite number of
+resources in the same directory.
+
+1. Create a base directory named `<book-id>`
+1. Create a file named `0=ocfl_object_1.0` with the appropriate contents.
+1. Create a v1 directory
+1. Copy all content files and metadata files into `<book-id>/v1/content`, separated
+   by ID. So, for instance, `<page1>` would be in `<book-id>/v1/content/<page1>`,
+   and have two files: `<page1>.json` and
+   `<page1>-<file-name>.<file-extension>`.
+1. Create an `inventory.json` in `v1`, as per the specification, with SHA512s of
+   every file in `content` as the keys of the `manifest` property.
+1. Create `inventory.json.sha512` in the `v1` directory.
+1. Copy `inventory.json` and `inventory.json.sha512` to the `<book-id>`
+   directory.
+
+**Pros**:
+
+1. We now have an OCFL repository which is theoretically readable by future OCFL
+   browsers, and will be understood by partner institutions.
+1. This structure just needs a file system to store itself and support
+   versioning, there's no need for an object store's feature-set.
+1. We can de-duplicate as part of generating the `inventory.json` if we'd like
+   to, using up less space than the other options.
+
+**Cons**:
+
+1. The inventory file acts as a pessimistic lock in the same way as BagIt's
+   manifest file.
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the inventory will be
+   corrupted by a race condition, and versions will be unable to be properly
+   calculated:
```

Comment:
A problem with many versioning systems is determining what constitutes a version. For example if my title lacks a period at the end of it and I add the period is that a new version? If I have a massive data set a researcher is working on do I really want to keep a new copy each time it changes? Or do I want to version it at moments that make sense in the researcher's lifecycle (after a publication, before major grant work, etc.). OCFL provides freedom for the institution to make those decisions. 

OCFL does not perscribe when a version is created, we leave that up to the software that sits on top of the storage root.

Reply:
I'm not sure how I can put that here. In our case we want every change post-complete to not be lost - every versioning system for a preservation store you're not using as your database has that feature right? It's just whenever you decide to preserve its state again. So I guess the race condition is more "prevent future versions from being created until this one is done"?

Reply:
Oh! Unless you're proposing it's kosher to just mutate the files in the existing files and fix up the checksums? I guess that'd be fine. Do you think it'd be helpful to put that here? It might make this case more complex, and I'm not sure we'd do it

Reply:
I think the implication is that we'd need to buffer changes and not mutate the OCFL package the way we are mutating our packages now.  I think that's a thing we should look into, but not something we need to resolve right now.

Reply:
I agree @escowles -- I think this is the key difference between BagIt/OCFL and your GCS-based approached. In BagIt/OCFL the object is very well-defined and bounded, there is no notion of a hierarchy of objects where something can be added inside without affected the data (manifest) for the whole. IMO this provides clarity for preservation but comes with a cost. Links from one object to other objects could provide a way around this but is hasn't been thought out for OCFL (maybe we should write up a use case for that)

Reply:
:+1: I look forward to seeing what OCFL clients do to deal with all this. It seems like the reasonable strategy is something less like "store this resource in a package" and more like "preserve this whole hierarchy please," with some pretty amazing logic for handling diffing with potentially very little access to the material upstream (in cases of expensive cloud storage - or even delayed storage with Glacier).